File: scripts/prepare-installer.py
# /// script
# requires-python = ">=3.10"
# ///

"""Prepare installer staging directory for `cargo wix`.

Called as the `before` hook by cargo-wix. Builds all crates (triggering
build.rs which builds v2ray-plugin and downloads wintun), then stages
the built binaries into a directory for WiX.

Environment variables (set by cargo-wix):
    WIX_TARGET_DIR     — Cargo target directory
    WIX_WORKSPACE_ROOT — Workspace root
"""

import os
import shutil
import subprocess
import sys
from pathlib import Path


def main() -> None:
    target_dir = Path(os.environ["WIX_TARGET_DIR"])
    release_dir = target_dir / "release"
    stage = release_dir / "installer-stage"

    # Build =====

    print("==> Building release binaries...", file=sys.stderr)
    result = subprocess.run(["cargo", "build", "--release", "--workspace"])
    if result.returncode != 0:
        print("cargo build failed", file=sys.stderr)
        sys.exit(1)

    # Stage =====

    stage.mkdir(parents=True, exist_ok=True)
    print(f"==> Staging installer files to {stage}", file=sys.stderr)

    # hole.exe — main binary
    link_or_copy(release_dir / "hole.exe", stage / "hole.exe")

    # v2ray-plugin.exe — built by crates/gui/build.rs
    v2ray_dir = Path(".cache/gui/v2ray-plugin")
    candidates = list(v2ray_dir.glob("v2ray-plugin-*.exe"))
    if len(candidates) == 0:
        print(f"error: no v2ray-plugin binary found in {v2ray_dir}", file=sys.stderr)
        sys.exit(1)
    if len(candidates) > 1:
        print(
            f"error: multiple v2ray-plugin binaries in {v2ray_dir}: {candidates}",
            file=sys.stderr,
        )
        sys.exit(1)
    link_or_copy(candidates[0], stage / "v2ray-plugin.exe")

    # wintun.dll — downloaded by crates/gui/build.rs
    link_or_copy(Path(".cache/gui/wintun/wintun.dll"), stage / "wintun.dll")

    print("==> Staging complete", file=sys.stderr)


def link_or_copy(src: Path, dst: Path) -> None:
    dst.unlink(missing_ok=True)
    try:
        os.link(src, dst)
        print(f"  {dst.name} (hardlinked)", file=sys.stderr)
    except OSError:
        shutil.copy2(src, dst)
        print(f"  {dst.name} (copied)", file=sys.stderr)


if __name__ == "__main__":
    main()

File: scripts/prepare-installer.rs
#!/usr/bin/env cargo
//! Prepare installer staging directory for `cargo wix`.
//!
//! This script is called as the `before` hook by cargo-wix. It:
//! 1. Runs `cargo build --release --workspace` (which triggers build.rs to
//!    build v2ray-plugin from Go source and download wintun.dll)
//! 2. Copies the built binaries into a staging directory for WiX

use std::path::{Path, PathBuf};
use std::process::Command;
use std::{env, fs};

fn main() {
    let target_dir = env_path("WIX_TARGET_DIR");
    let release_dir = target_dir.join("release");
    let stage = release_dir.join("installer-stage");

    // Build =====

    eprintln!("==> Building release binaries...");
    let status = Command::new("cargo")
        .args(["build", "--release", "--workspace"])
        .status()
        .expect("failed to run cargo build");
    if !status.success() {
        eprintln!("cargo build failed");
        std::process::exit(1);
    }

    // Stage =====

    fs::create_dir_all(&stage).expect("failed to create staging directory");
    eprintln!("==> Staging installer files to {}", stage.display());

    // hole.exe — main binary
    hardlink_or_copy(&release_dir.join("hole.exe"), &stage.join("hole.exe"));

    // v2ray-plugin.exe — built by crates/gui/build.rs into .cache/gui/v2ray-plugin/
    let v2ray_src = find_single_file(
        &PathBuf::from(".cache/gui/v2ray-plugin"),
        "v2ray-plugin-",
        ".exe",
    );
    hardlink_or_copy(&v2ray_src, &stage.join("v2ray-plugin.exe"));

    // wintun.dll — downloaded by crates/gui/build.rs into .cache/gui/wintun/
    hardlink_or_copy(
        &PathBuf::from(".cache/gui/wintun/wintun.dll"),
        &stage.join("wintun.dll"),
    );

    eprintln!("==> Staging complete");
}

fn env_path(var: &str) -> PathBuf {
    PathBuf::from(env::var(var).unwrap_or_else(|_| panic!("{var} not set")))
}

/// Find the single file in `dir` matching `prefix*suffix`.
fn find_single_file(dir: &Path, prefix: &str, suffix: &str) -> PathBuf {
    let entries: Vec<PathBuf> = fs::read_dir(dir)
        .unwrap_or_else(|e| panic!("failed to read {}: {e}", dir.display()))
        .filter_map(|e| e.ok())
        .map(|e| e.path())
        .filter(|p| {
            p.is_file()
                && p.file_name()
                    .and_then(|n| n.to_str())
                    .map(|n| n.starts_with(prefix) && n.ends_with(suffix))
                    .unwrap_or(false)
        })
        .collect();

    match entries.len() {
        0 => panic!("no file matching {prefix}*{suffix} in {}", dir.display()),
        1 => entries.into_iter().next().unwrap(),
        n => panic!(
            "expected 1 file matching {prefix}*{suffix} in {}, found {n}",
            dir.display()
        ),
    }
}

/// Try to hardlink, fall back to copy.
fn hardlink_or_copy(src: &Path, dst: &Path) {
    let _ = fs::remove_file(dst);

    if fs::hard_link(src, dst).is_ok() {
        eprintln!(
            "  {} (hardlinked)",
            dst.file_name().unwrap().to_string_lossy()
        );
    } else {
        fs::copy(src, dst).unwrap_or_else(|e| {
            panic!(
                "failed to copy {} -> {}: {e}",
                src.display(),
                dst.display()
            )
        });
        eprintln!("  {} (copied)", dst.file_name().unwrap().to_string_lossy());
    }
}
